draft-west-cookie-incrementalism.md

```diff
@@ -401,85 +404,304 @@ from "have the same registrable domain" (or "is an exact match for") to say
+{{RFC6265bis}} points out. Similarly to above, attackers operating on a
+different port could influence otherwise secured traffic.
+
+We should remedy this defect by storing both a "scheme" and "port"
+component along with the cookie, and use those components in cookies'
+matching algorithms to ensure that a cookie is only sent to the origin
+that originally set it, thus keeping origins' state separate by default.
+
+It's possible that a site operator purposefully wants their cookies to
+be usable across ports. For these situations it seems reasonable to
+allow a server to specify that a given cookie should be accessible by
+different origins. Since the Domain attribute already does this, by
+matching multiple hosts, we can extend it to also match all ports.
+Thereby allowing a cookie with the Domain attribute, a "domain cookie",
+to be accessible from any port value. This behavior extends to
+overwriting cookies as well. Examples:
 
-We should remedy this defect by storing a `scheme` component along with the cookie, and using that
-component in cookies' matching algorithms to ensure that secure and non-secure origins' state is
-clearly distinguishable and separate. This is accomplished as follows:
+1. https://example.com:123 sets a cookie: `Set-Cookie: a=domaincookie;
+    Domain=example.com`. The user then visits https://example.com:456
+    which sets the same cookie. This second cookie would match the
+    first and then overwrite it. 
+2. https://example.com:123 sets a cookie `Set-Cookie: a=domaincookie;
+    Domain=example.com`. The user then visits https://example.com:456
+    which sets `Set-Cookie: a=origincookie`. The second cookie would
+    match the first and overwrite it.
 
-First, alter the Storage Model defined in Section 5.4 of {{RFC6265bis}} by adding `scheme` to the
-list of fields the user agent stores about each cookie, and setting it when creating a cookie by
-altering step 2 of the same algorithm from:
+Given that these "domain cookies" are less trusted (due to their
+accessibility by different hosts and ports) we should try and prevent
+them from shadowing "origin cookies". Example:
+
+1. https://example.com:456 sets a cookie `Set-Cookie: a=origincookie`.
+   The user then visits https://example.com:123 which tries to set a
+   cookie `Set-Cookie: a=domaincookie; Domain=example.com`. The second
+   cookie would fail to be set because the cookie it's attempting to
+   overwrite is an origin cookie.
+
+Finally, we don't ever want to allow a cookie to pass between schemes,
+given the huge security differences between them. So there should be no
+way for a server to specify that a given cookie should be sent to a
+different scheme.
+
+How we accomplish this is as follows:
+
+First, add the concept of port matching which helps to simplify checking
+if a cookie would match a port value. We can do that by adding a new
+section under 5.1 (this new section depends on the modification to 5.4
+below)
```

Comment:
Are you recommending to split up the introduction of these concepts within the explanation or to ship the feature in two distinct parts?

Reply:
It's been a little while, but I think I was suggesting treating the `scheme` and the `port` changes as distinct phases that you could ship along different timelines.

Reply:
(Reading through this again, I still think these changes would be simpler to introduce as distinct alterations. Perhaps it would make sense to chat through your deployment story, as I might well be missing something that makes it prudent to ship all this as one change.)

Reply:
Doing as part of one proposal was primarily for three reasons:

- The story of aligning cookies on the origin, a well known concept (`Domain` not withstanding).
- The working assumption is that binding to scheme will be harder to ship, and that adding port onto that isn't much more work
  - Metrics also indicate that restriction the port is likely to be less breaking than scheme
- Avoiding annoying users more than necessary via death by 1000 changes.

Reply:
I was thinking about this in the other direction: `port` seems much easier to ship, so we shouldn't block it on `scheme`. :)

Reply:
Spoke offline, I'm still considering shipping the feature separately but I won't block the spec change work on that.